torch/csrc/jit/codegen/cuda/iter_visitor.cpp

```diff
@@ -31,125 +28,250 @@ std::vector<Statement*> IterVisitor::next(Expr* expr) {
   return {expr->inputs().begin(), expr->inputs().end()};
 }
 
+// Remove any stmt in stmts that is in visited
+namespace {
+void remove_visited(
+    std::vector<Statement*>& stmts,
+    const std::unordered_set<Statement*>& visited) {
+  std::deque<std::vector<Statement*>::iterator> to_erase;
+  for (auto it = stmts.begin(); it != stmts.end(); it++) {
+    if (visited.find(*it) != visited.end())
+      to_erase.push_back(it);
+  }
+
+  while (!to_erase.empty()) {
+    stmts.erase(to_erase.back());
+    to_erase.pop_back();
+  }
+}
+} // namespace
+
 void IterVisitor::traverseFrom(
     Fusion* const fusion,
-    const std::vector<Val*>& from) {
+    const std::vector<Val*>& from,
+    bool traverseAllPaths) {
   FusionGuard fg(fusion);
+  std::unordered_set<Statement*> visited;
+  stmt_stack.clear();
+  stmt_stack.push_back(std::vector<Statement*>(from.rbegin(), from.rend()));
 
-  std::set<Statement*> visited;
-  std::deque<Statement*> to_visit;
-
-  std::queue<Val*> outputs_to_visit;
-  for (Val* entry : from)
-    outputs_to_visit.emplace(entry);
-
-  while (!outputs_to_visit.empty()) {
-    if (stopCondition())
-      break;
-
-    to_visit.push_front(outputs_to_visit.front());
-    outputs_to_visit.pop();
-    while (!to_visit.empty()) {
-      if (stopCondition())
-        break;
-
-      Statement* stmt = to_visit.front();
-      std::vector<Statement*> inps = next(stmt);
-      for (auto it = inps.rbegin(); it != inps.rend(); it++) {
-        Statement* inp = *it;
-        if (visited.find(inp) == visited.end()) {
-          toVisitCallback(inp);
-          to_visit.emplace_front(inp);
-        }
-      }
+  while (!stmt_stack.empty()) {
+    auto next_stmts = next(stmt_stack.back().back());
 
-      if (to_visit.front() != stmt) {
-        continue;
-      }
+    // Remove statements we already visited if we're not traversing all paths
+    if (!traverseAllPaths)
+      remove_visited(next_stmts, visited);
 
-      to_visit.pop_front();
-      if (visited.find(stmt) == visited.end()) {
-        handle(stmt);
-        visited.emplace(stmt);
+    // Traverse down until we get to a leaf
+    while (!next_stmts.empty()) {
+      stmt_stack.push_back(
```

Comment:
nitpick: `emplace_back`

Reply:
Why emplace_back?

Reply:
stmt_stack is poorly named, but is a vector. Maybe you were thinking it should be top?

Reply:
sorry for not being clear. This is just a nitpick on styling & shorter code.

I was suggesting `stmt_stack.emplace_back(next_stmts.rbegin(), next_stmts.rend());` instead of `stmt_stack.push_back(std::vector<Statement*>(next_stmts.rbegin(), next_stmts.rend()));`